src/group.cpp

```diff
@@ -425,6 +425,7 @@ void Group::Generate(IdList<Entity,hEntity> *entity,
             // Get some arbitrary point in the sketch, that will be used
             // as a reference when defining top and bottom faces.
             hEntity pt = { 0 };
+            // Not using range-for here because we're changing the size of entity in the loop.
```

Comment:
Oh *of course* that's why my own range-for patch (in a local branch that got buried years ago) didn't work. Iterator invalidation, the second worst thing in C++ that "modern" C++ does absolutely nothing to help with...

Reply:
Can we add iterators that don't get invalidated when inserting into the collection, maybe? Store an index as well as pointer to the element.

Reply:
maybe as "future work". I did have some patches that make an IdListAdditions type, which basically queues up things to add so it can be bulk-added at the end via a merge. Would also avoid invalidating iterators if iterating an idlist, but not sure the clarity tradeoff is worth it. (I reverted the patch on my dev branch after doing the port to std::vector as a backend, since that gave the equivalent performance improvement. Think it was allocation behavior that was gettting me.)

Reply:
Here's that branch rebased on this, and trimmed. https://github.com/rpavlik/solvespace/commits/idlist-additions

See in particular this: https://github.com/rpavlik/solvespace/commit/93f5e410e586281e641f52056d409de2dd81263d  I only did this in the few places where I saw it coming up a lot on profiling data, but I think this could be used to avoid iterator invalidation in loops as well - instead of adding directly, add to an "additions" object, then "MoveIntoList" after iteration is complete.